**Put any package imports you need in the below space**

In [2]:
#import matplotlib inline # you will want this if plotting with pyplot
%matplotlib inline

import matplotlib
import numpy as np
import matplotlib.pyplot as plt

import networkx as nx

**Exercise 1**: Generating the adjacency matrix A for graphs of particular type. Write a script that generates the adjacency matrix A for each of the following graphs and prints this matrix onto the screen:
1. K5: 5-clique, or a fully connected simple, undirected graph of 5 nodes
2. K5.3: two disconnected components, C1 and C2, where C1 is a 5-clique K5 and C2 is a 3-clique K3
3. K5.3e: Almost the same as K53 but there is a single edge connecting the two components
4. B2.3: Complete bi-partite graph with n1=2 nodes in the first part and n2=3 nodes in the second part
5. S5: A 5-vertex star (one central "hub" node that connects to all the other "spoke" nodes)
6. P5: A simple path of 5 vertices


In [3]:
print("K5 - 5-clique")
k_5 = nx.complete_graph(5)
# nx.draw(k_5)
A = nx.adjacency_matrix(k_5)
print(A.todense())

print("K5.3 - C1 is K5, C2 is K3")
k_3 = nx.complete_graph(3)
k_5_3 = nx.disjoint_union(k_5, k_3)
# nx.draw(k_5_3)
B = nx.adjacency_matrix(k_5_3)
print(B.todense())

print("K5.3e - Edge between C1 and C2 in K5.3")
k_5_3e = nx.disjoint_union(k_5, k_3)
k_5_3e.add_edge(1,6)
# nx.draw(k_5_3e)
C = nx.adjacency_matrix(k_5_3e)
print(C.todense())

print("B2.3 - Bi-partite graph with n1=2 and n2=3")
b_2_3 = nx.complete_bipartite_graph(2,3)
# nx.draw(b_2_3)
D = nx.adjacency_matrix(b_2_3)
print(D.todense())

print("S5 - 5-vertex star")
s_5 = nx.star_graph(5)
# nx.draw(s_5)
E = nx.adjacency_matrix(s_5)
print(E.todense())

print("P5 - Simple path with 5 vertices")
p_5 = nx.path_graph(5)
# nx.draw(p_5)
F = nx.adjacency_matrix(p_5)
print(F.todense())

def print_adjacency_matrix(graph):
    print(nx.adjacency_matrix(graph).todense())
    
graphs = [k_5, k_5_3, k_5_3e, b_2_3, s_5, p_5]

K5 - 5-clique
[[0 1 1 1 1]
 [1 0 1 1 1]
 [1 1 0 1 1]
 [1 1 1 0 1]
 [1 1 1 1 0]]
K5.3 - C1 is K5, C2 is K3
[[0 1 1 1 1 0 0 0]
 [1 0 1 1 1 0 0 0]
 [1 1 0 1 1 0 0 0]
 [1 1 1 0 1 0 0 0]
 [1 1 1 1 0 0 0 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 0 1 0 1]
 [0 0 0 0 0 1 1 0]]
K5.3e - Edge between C1 and C2 in K5.3
[[0 1 1 1 1 0 0 0]
 [1 0 1 1 1 0 1 0]
 [1 1 0 1 1 0 0 0]
 [1 1 1 0 1 0 0 0]
 [1 1 1 1 0 0 0 0]
 [0 0 0 0 0 0 1 1]
 [0 1 0 0 0 1 0 1]
 [0 0 0 0 0 1 1 0]]
B2.3 - Bi-partite graph with n1=2 and n2=3
[[0 0 1 1 1]
 [0 0 1 1 1]
 [1 1 0 0 0]
 [1 1 0 0 0]
 [1 1 0 0 0]]
S5 - 5-vertex star
[[0 1 1 1 1 1]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]]
P5 - Simple path with 5 vertices
[[0 1 0 0 0]
 [1 0 1 0 0]
 [0 1 0 1 0]
 [0 0 1 0 1]
 [0 0 0 1 0]]


**Exercise 2**: Generating the degree matrix D for a given adjacency matrix A. Write a script that generates a degree matrix for each of the adjacency matrices in Exercise 1. Note that the degree matrix is a diagonal matrix where all the positions except for the diagonal are zero's. The diagonal elements correspond to the degrees of the corresponding nodes, namely Dii = degree(v_i).

In [4]:
def generate_degree_matrix(graph):
    no_of_nodes = graph.number_of_nodes()
    node_degrees = list(graph.degree(graph.nodes()).values())
    node_degree_array = np.asarray(node_degrees)
    identity_matrix = np.eye(no_of_nodes).astype(int)
    np.fill_diagonal(identity_matrix, node_degree_array)
    print(identity_matrix)
    return identity_matrix
'''
deg_k5 = generate_degree_matrix(k_5)
deg_k53 = generate_degree_matrix(k_5_3)
deg_k53e = generate_degree_matrix(k_5_3e)
deg_b23 = generate_degree_matrix(b_2_3)
deg_s5 = generate_degree_matrix(s_5)
deg_p5 = generate_degree_matrix(p_5)
'''

degree_matrices = [generate_degree_matrix(g) for g in graphs]

[[4 0 0 0 0]
 [0 4 0 0 0]
 [0 0 4 0 0]
 [0 0 0 4 0]
 [0 0 0 0 4]]
[[4 0 0 0 0 0 0 0]
 [0 4 0 0 0 0 0 0]
 [0 0 4 0 0 0 0 0]
 [0 0 0 4 0 0 0 0]
 [0 0 0 0 4 0 0 0]
 [0 0 0 0 0 2 0 0]
 [0 0 0 0 0 0 2 0]
 [0 0 0 0 0 0 0 2]]
[[4 0 0 0 0 0 0 0]
 [0 5 0 0 0 0 0 0]
 [0 0 4 0 0 0 0 0]
 [0 0 0 4 0 0 0 0]
 [0 0 0 0 4 0 0 0]
 [0 0 0 0 0 2 0 0]
 [0 0 0 0 0 0 3 0]
 [0 0 0 0 0 0 0 2]]
[[3 0 0 0 0]
 [0 3 0 0 0]
 [0 0 2 0 0]
 [0 0 0 2 0]
 [0 0 0 0 2]]
[[5 0 0 0 0 0]
 [0 1 0 0 0 0]
 [0 0 1 0 0 0]
 [0 0 0 1 0 0]
 [0 0 0 0 1 0]
 [0 0 0 0 0 1]]
[[1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 2 0 0]
 [0 0 0 2 0]
 [0 0 0 0 1]]


**Exercise 3**: Generating the graph Laplacian matrix L for a given adjacency matrix A and its degree matrix D. Write a script that generates the graph Laplacian matrix L = D - A for each of the adjacency matrices in Exercise 1.

In [5]:
def laplacian_matrix(graph):
    temp = nx.laplacian_matrix(graph).todense()
    print(temp)
    return temp

laplacian_matrices = [laplacian_matrix(g) for g in graphs]

[[ 4 -1 -1 -1 -1]
 [-1  4 -1 -1 -1]
 [-1 -1  4 -1 -1]
 [-1 -1 -1  4 -1]
 [-1 -1 -1 -1  4]]
[[ 4 -1 -1 -1 -1  0  0  0]
 [-1  4 -1 -1 -1  0  0  0]
 [-1 -1  4 -1 -1  0  0  0]
 [-1 -1 -1  4 -1  0  0  0]
 [-1 -1 -1 -1  4  0  0  0]
 [ 0  0  0  0  0  2 -1 -1]
 [ 0  0  0  0  0 -1  2 -1]
 [ 0  0  0  0  0 -1 -1  2]]
[[ 4 -1 -1 -1 -1  0  0  0]
 [-1  5 -1 -1 -1  0 -1  0]
 [-1 -1  4 -1 -1  0  0  0]
 [-1 -1 -1  4 -1  0  0  0]
 [-1 -1 -1 -1  4  0  0  0]
 [ 0  0  0  0  0  2 -1 -1]
 [ 0 -1  0  0  0 -1  3 -1]
 [ 0  0  0  0  0 -1 -1  2]]
[[ 3  0 -1 -1 -1]
 [ 0  3 -1 -1 -1]
 [-1 -1  2  0  0]
 [-1 -1  0  2  0]
 [-1 -1  0  0  2]]
[[ 5 -1 -1 -1 -1 -1]
 [-1  1  0  0  0  0]
 [-1  0  1  0  0  0]
 [-1  0  0  1  0  0]
 [-1  0  0  0  1  0]
 [-1  0  0  0  0  1]]
[[ 1 -1  0  0  0]
 [-1  2 -1  0  0]
 [ 0 -1  2 -1  0]
 [ 0  0 -1  2 -1]
 [ 0  0  0 -1  1]]


Answer the following questions:
1. Is L a sparse matrix?
2. In what positions does L have non-zero elements?
3. What are the values of the non-diagonal and non-zero elements?
4. What does L contain along its diagonal?



**Answers**:
1. No. L is not a sparse matrix.
2. Diagonal, and positions where there was an edge in corresponding adjacency matrix
3. Negative
4. Degrees of vertices

**Exercise 4**: Generating the graph spectrum, or the multiset of the eigenvalues of the graph adjacency matrix A. Write a script that calculates the eigenvalues of the graph adjacency matrix for each of the matrices in Exercise 1. Plot the eigenvalues in the increasing order of their values

In [6]:
def generate_eigenvalues_adj(graph):
    eigen_adj = nx.adjacency_spectrum(graph)
    eigen_adj_rounded = [round(x,5) for x in eigen_adj.real]
    eigen_adj_rounded.sort()
    print(eigen_adj_rounded)
    return eigen_adj_rounded

eigenvalues_adjcency = [generate_eigenvalues_adj(g) for g in graphs]

[-1.0, -1.0, -1.0, -1.0, 4.0]
[-1.0, -1.0, -1.0, -1.0, -1.0, -1.0, 2.0, 4.0]
[-1.7798700000000001, -1.0, -1.0, -1.0, -1.0, -0.33659, 2.0549200000000001, 4.0615300000000003]
[-2.4494899999999999, 0.0, 0.0, 0.0, 2.4494899999999999]
[-2.2360699999999998, 0.0, 0.0, 0.0, 0.0, 2.2360699999999998]
[-1.7320500000000001, -1.0, -0.0, 1.0, 1.7320500000000001]


Answer the following questions:
1. What can you say about the eigenvalues of the complete graph (K5): the number of unique eigenvalues, the largest and the smallest eigenvalues, the multiplicity (how many times the same eigenvalue appears) of each eigenvalue?
2. What is the graph spectrum of the bi-partite graph, B2.3? If n1 = n and n2 = m (a general complete bi-partite graph), then what can you say about its graph spectrum? [Hint: check sqrt(n * m)] If \lambda is the eigenvalue of the bi-partite graph, will minus \lambda be also the eigenvalue?
3. What is the largest eigenvalue of the star graph S5? If S5 were generalized to an N-vertex star, what could you say about the value of its largest eigenvalue?
4. What is the largest eigenvalue of the path graph P5? As the length of the path increases, what can you say about the changes in the largest eigenvalue?
5. How does the largest eigenvalue of the path P5 (or its more generalization to an arbitrary length) compare with the largest eigenvalues of the star graph or the complete graph? If you are asked to sort the largest eigenvalue of the path, the star, and the clique) in the increasing order, what kind of relationship would you assign (E.g., \lambda{path} > or < than \lambda{star})?


**Answers**:

1. Excluding the largest eigenvalue, which equals to +4.0, other 4 eigenvalues equal to -1.0.
2. The extreme eigenvalues equal to *sqrt(n * m)*, and differ in sign. The rest of the values are zero. If \lambda is the eigenvalue of the bi-partite graph, minus \lambda be also an eigenvalue.
3. The largest eigenvalue for the star graph *S5* is 2.236, which is *sqrt(5)*. Generalizing for an N-vertex star, we can say that largest eigenvalue = *sqrt(N)*.
4. The largest eigenvalue for the path graph *P5* is 1.732, which is *sqrt(3)*. That *'3'* accounts for the nodes in-between the path (excluding the extreme nodes from path or considering nodes with degree 2 = N-2). Hence, as the path length would increase, so would the largest eigenvalue.
5. \lambda{path} < \lambda{star} < \lambda{clique} for the same number of nodes N.

**Exercise 5**: Generating the graph spectrum, or the multiset of the eigenvalues of the graph Laplacian. Write a script that calculates the eigenvalues of the graph Laplacian for each of the graphs in Exercise 1. Plot the eigenvalues in the increasing order of their values. 

In [7]:
def generate_eigenvalues_laplacian(graph):
    eigen_lapl = nx.laplacian_spectrum(graph)
    eigen_lapl_rounded = [round(x,5) for x in eigen_lapl.real]
    eigen_lapl_rounded.sort()
    print(eigen_lapl_rounded)
    return eigen_lapl_rounded

eigenvalues_laplacian = [generate_eigenvalues_laplacian(g) for g in graphs]

[-0.0, 5.0, 5.0, 5.0, 5.0]
[-0.0, -0.0, 3.0, 3.0, 5.0, 5.0, 5.0, 5.0]
[0.0, 0.37380000000000002, 3.0, 3.4848599999999998, 5.0, 5.0, 5.0, 6.1413399999999996]
[-0.0, 2.0, 2.0, 3.0, 5.0]
[-0.0, 1.0, 1.0, 1.0, 1.0, 6.0]
[0.0, 0.38196999999999998, 1.3819699999999999, 2.6180300000000001, 3.6180300000000001]


Answer the following questions:
1. What can you say about the largest and the smallest eigenvalues?
2. What is the multiplicity (how many times the same eigenvalue appears) of the zero eigenvalue for each of the cases?
3. If K53 graph would be generalized to include c>2 components, what can you say about the multiplicity of the zero eigenvalues?
4. If graph G is connected (i.e., the number of disconnected components is one), what can you say about the multiplicity of the zero eigenvalue?
5. For the bi-partite graph, what is the value of the second smallest eigenvalue?
6. Is the vector, whose components consist of 1's only, the eigenvector of the Laplacian? If it is, then what is its corresponding eigenvalue?
7. Suppose the graph Laplacian matrix has the zero eigenvalue of multiplicity k. Can you say anything about the structure of such a graph?


**Answers:**

1. Smallest eigenvalues are approximately zero. I do not observe a pattern in largest eigenvalues.
2. Except for the second graph, which is a graph with two disconnected components, cliques of sizes 5 and 3, for every other graph, the multiplicity of zero eigenvalues is one.
3. In such a case, the multiplicity of zero eigenvalues would be equal to number of disconnected components present in the graph.
4. If the graph is connected, the multiplicity of zero eigenvalues is one.
5. For the bi-partite graph, the second smallest eigenvalue is 3.0
6. 
7. Such a graph has k- disconnected components.